controller/lib/pid/pid.cpp

```diff
@@ -18,238 +17,90 @@ limitations under the License.
  **********************************************************************************************/
 
 #include "pid.h"
+#include "algorithm.h"
 #include "hal.h"
 
-/*Constructor (...)*********************************************************
- *    The parameters specified here are those for for which we can't set up
- *    reliable defaults, so we need to have the user set them.
- ***************************************************************************/
-PID::PID(float *Input, float *Output, float *Setpoint, float Kp, float Ki,
-         float Kd, bool POnE, bool DOnE, int ControllerDirection)
-    : SampleTime(milliseconds(100)), nextSampleTime(Hal.now()),
-      lastUpdateTime(Hal.now() - SampleTime) {
-  myOutput = Output;
-  myInput = Input;
-  mySetpoint = Setpoint;
-  inAuto = false;
-
-  // default output limit corresponds to the arduino pwm limits
-  PID::SetOutputLimits(0, 255);
-
-  PID::SetControllerDirection(ControllerDirection);
-  PID::SetTunings(Kp, Ki, Kd, POnE, DOnE);
-}
-
-/*Constructor (...)*********************************************************
- *    To allow backwards compatability for v1.1, or for people that just want
- *    to use Proportional on Error without explicitly saying so
- ***************************************************************************/
+PID::PID(float kp, float ki, float kd, ProportionalTerm p_term,
+         DifferentialTerm d_term, ControlDirection direction)
+    : kp_(kp), ki_(ki), kd_(kd), p_term_(p_term), d_term_(d_term),
+      direction_(direction) {}
 
-PID::PID(float *Input, float *Output, float *Setpoint, float Kp, float Ki,
-         float Kd, int ControllerDirection)
-    : PID::PID(Input, Output, Setpoint, Kp, Ki, Kd, P_ON_E, D_ON_M,
-               ControllerDirection) {}
-
-/* Compute()
- *********************************************************************** This,
- *as they say, is where the magic happens.  this function should be called every
- *time "void loop()" executes.  the function will decide for itself whether a
- *new pid Output needs to be computed.  returns true when the output is
- *computed, false when nothing has been done.
- **********************************************************************************/
-bool PID::Compute() {
-  if (!inAuto)
-    return false;
+bool PID::Compute(float input, float setpoint, float *output) {
   Time now = Hal.now();
+  if (!initialized_) {
+    last_input_ = input;
+    last_error_ = setpoint - input;
+    output_sum_ = stl::clamp(*output, out_min_, out_max_);
```

Comment:
When we do

```
  float output;
  myPID.Compute(/*input=*/cur_pressure.kPa(),
       /*setpoint=*/desired_state.setpoint_pressure.kPa(), &output);
```

Isn't output used uninitialized here?

(Could we fix this bug by returning a float instead of a boolean?)

Reply:
Whew, good catch! Too bad testing with msan didn't work out... Fixed.